pkg/operator/ceph/cluster/crash/crash.go

```diff
@@ -122,57 +125,121 @@ func (r *ReconcileNode) createOrUpdateCephCrash(node corev1.Node, tolerations []
+	// minimum k8s version required for v1 cronJob is 'v1.21.0'. Apply v1 if k8s version is at least 'v1.21.0', else apply v1beta1 cronJob.
+	if mink8sVersion {
+		cronJob := createV1CronJob(cephCluster.GetNamespace())
+		err := controllerutil.SetControllerReference(&cephCluster, cronJob, r.scheme)
+		if err != nil {
+			return controllerutil.OperationResultNone, errors.Errorf("failed to set owner reference of deployment %q", cronJob.Name)
+		}
 
-	mutateFunc := func() error {
+		// Adding volumes to pods containing data needed to connect to the ceph cluster.
+		volumes := controller.DaemonVolumesBase(config.NewDatalessDaemonDataPathMap(cephCluster.GetNamespace(), cephCluster.Spec.DataDirHostPath), "")
+		volumes = append(volumes, keyring.Volume().CrashCollector())
 
-		// labels for the pod, the deployment, and the deploymentSelector
-		cronJobLabels := map[string]string{
-			k8sutil.AppAttr: prunerName,
+		mutateFunc := func() error {
+
+			// labels for the pod, the deployment, and the deploymentSelector
+			cronJobLabels := map[string]string{
+				k8sutil.AppAttr: prunerName,
+			}
+			cronJobLabels[k8sutil.ClusterAttr] = cephCluster.GetNamespace()
+
+			cronJob.ObjectMeta.Labels = cronJobLabels
+			cronJob.Spec.JobTemplate.Spec.Template = corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: cronJobLabels,
+				},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						getCrashPruneContainer(cephCluster, *cephVersion),
+					},
+					RestartPolicy: corev1.RestartPolicyNever,
+					HostNetwork:   cephCluster.Spec.Network.IsHost(),
+					Volumes:       volumes,
+				},
+			}
+
+			cronJob.Spec.Schedule = pruneSchedule
+			// After 100 failures, the cron job will no longer run.
+			// To avoid this, the cronjob is configured to only count the failures
+			// that occurred in the last hour.
+			deadline := int64(60)
+			cronJob.Spec.StartingDeadlineSeconds = &deadline
+
+			return nil
 		}
-		cronJobLabels[k8sutil.ClusterAttr] = cephCluster.GetNamespace()
 
-		cronJob.ObjectMeta.Labels = cronJobLabels
-		cronJob.Spec.JobTemplate.Spec.Template = corev1.PodTemplateSpec{
-			ObjectMeta: metav1.ObjectMeta{
-				Labels: cronJobLabels,
-			},
-			Spec: corev1.PodSpec{
-				Containers: []corev1.Container{
-					getCrashPruneContainer(cephCluster, *cephVersion),
+		return controllerutil.CreateOrUpdate(context.TODO(), r.client, cronJob, mutateFunc)
+	} else {
+		cronJob := createV1Beta1CronJob(cephCluster.GetNamespace())
+		err := controllerutil.SetControllerReference(&cephCluster, cronJob, r.scheme)
+		if err != nil {
+			return controllerutil.OperationResultNone, errors.Errorf("failed to set owner reference of deployment %q", cronJob.Name)
+		}
+
+		// Adding volumes to pods containing data needed to connect to the ceph cluster.
+		volumes := controller.DaemonVolumesBase(config.NewDatalessDaemonDataPathMap(cephCluster.GetNamespace(), cephCluster.Spec.DataDirHostPath), "")
```

Comment:
I think we could refactor this more to reduce duplication. For example, the volumes, mutateFunc, and other properties all look the same, right? Is it just the v1 or v1beta1 type that is different? What about structuring the code like this:
1. Initalize ObjectMeta, volumes, and mutateFunc in local vars
1. If K8s >= 1.21, create the v1 cron job and assign those local vars from step 1
1. If K8s < 1.21, create the v1beta1 cron job and assign those local vars from step 1

Reply:
restructured.